### ISSUE TITLE
Works only with "Latest call only" flagged

### ISSUE THREAD

**Issue by bbird81** [niekvlessert/audiocodes_syslog_tool#2]
Hi there! This project doesn't look mantained (4yrs since last commit), but I'll try to open an issue anyway...

Installed the whole thing on Ubuntu 22.04.1, running php8.1 (so probably that's already ad issue per se...), it comes up, syslog dumps data into postgres and sip calls are found when searched: life seems good.

Unfortunately if I click on a call a new page opens with URL ast/index.php?SID=58ddf8:18:7113449&device=sbc1 and in console I get the errors:
![immagine](https://user-images.githubusercontent.com/25413624/221933384-4c5d594a-9ea5-4dbd-9011-ae17b73d85b7.png)
Nothing is shown in this case, just the form with nothing below the line.

If I check the "Latest call only" flag instead, I get something:
![immagine](https://user-images.githubusercontent.com/25413624/221933743-07e73462-5f94-42c2-a7d0-338a88014724.png)

And below the messages related to that call, but only text.

My wild guess is that a graphic ladder should be there, because of this error in console:
![immagine](https://user-images.githubusercontent.com/25413624/221934912-b57c4481-38e3-4fda-b371-96c3100932e2.png)
I might be wrong on this and put too much expectations on this :)

So it doesn't really seem to work but it's almost there, my wild guess is that the newer version of PHP breaks something.

I will work on the problem on my own (even though I'm not a PHP developer) but if you have any suggestions it's more than welcome!

Thanks for your time and effort :)

**Comment by niekvlessert** [comment]
Hi,

Nice to see that someone is trying to use this. :)

Not sure what happens at this point, however I don't think it's PHP
related, the console is showing Javascript errors. What can happen of
course is that the syntax of the syslog changed a little bit and the code
can't handle that.

If you can send me the syslog file you are trying to parse I'm willing to
take a look.

Regards,

Niek

**Comment by bbird81** [comment]
Hi :) Thanks a lot for answering! I'm honestly quite surprised your project doesn't have like **A LOT** of stars and is not followed by many people: these AudioCodes devices become very popular and common in the last couple years because of MS Teams direct routing compatibility and ease of use/configuration...

Your project seems like the perfect low-budget (there more than a few products that do this, like AudioCodes OVOC, but they are pricey) answer to a very common problem, which is to archive SIP logs in order to troubleshoot past calls.

That said, I'm more than willing to share some logs and also help (even though I'm not a PHP developer unfortunately, but you know I might try to understand and maybe learn something :) ), just give me time to collect and redact them a little bit (nothing too secret, it's a lab environment... but better safe than sorry ;) ).

I'll attach them here soon, ok?

**Comment by bbird81** [comment]
[AudioCodes_logs.txt](https://github.com/niekvlessert/audiocodes_syslog_tool/files/10892605/AudioCodes_logs.txt)

Here's the logs you requested, let me know if I can help in any way!

Thanks again! :)

-t

**Comment by niekvlessert** [comment]
I'm pretty sure it has something to do with the syntax, I remember doing some nasty manipulating on that, but I forgot the details, it's been a few years.

I've been putting the syslog data in the database manually, but the stuff happening in the syslog is hard to emulate by hand. Could you please do a pg_dump on the whole database? You can do search/replace as well in the result as well like you did the in the syslog file.

Currently I'm not doing anything with VoIP anymore at work, but they've started using OVOC. However OVOC just collects the SIP data afaik, not the logic happening in the Audiocodes outside SIP, which is in the syslog, so in some cases you still have to go there. I agree that the tool could help people a lot. With the included maintenance stuff it can handle a lot of syslog. 

Probably if I added a screenshot of the SIP flow screenshot my project would have been a bit more popular. ;)

<img width="382" alt="image" src="https://user-images.githubusercontent.com/1013068/223574052-aaa60dde-217f-4183-9008-4a96715ec491.png">

**Comment by niekvlessert** [comment]
I've been debugging my own code some more. What happens is that the SBC also includes the CDR in the log. This happens at a different rsyslog facility, local1, the rest is of type local0. The rsylog configuration puts that data in a different table with the _cdr suffix. On that table is a postgres trigger enabled, which occurs when a database insert is done. If that insert contains CALL_END (the CDR record with the most data in it) in the message, then the message will be splitted and put in a another table with _cdr_formatted in as a suffix. Records found in there will show up if you don't enable the last_call_only field in the GUI. A manual emulation of that insert action with your data worked properly. So I think my database is now good enough to do some more debugging...

<img width="632" alt="image" src="https://user-images.githubusercontent.com/1013068/224153722-18a54379-624b-44b9-acc1-991307a33995.png">

**Comment by bbird81** [comment]
Ok so you don't need me to do the db dump anymore? That's great.

So for the search thing: if you got it right, it might be some problems in my syslog configuration after all, which doesn't trigger the cdr db insertion... could that be? I'll look into that just to be sure, 'cause judging from your screenshots seems like I might have not installed it correctly.

BTW yes I think that screenshot of the sip ladder and a little bit more explanation would have made your project hugely popular: that's what most of the customers of these appliances are looking for anyway. Logging, storing and retrieving them easily, with a sip ladder rendering: it covers 90% of use cases of customers.

I think there could easily be a market for this of product (if it becomes an out-of-the-box experience, like "install this package" or "execute this container")... think about it.

**Comment by bbird81** [comment]
I was too curious so I checked immediately :)

CALL_END entries are there in table _formatted:
![immagine](https://user-images.githubusercontent.com/25413624/224478021-df0d0371-03e3-4b80-81e3-81da1bdce906.png)

Let me know if you have any luck in correcting it or if I can be of any help in discussing the SIP scenario.

Thanks!

**Comment by bbird81** [comment]
Hi! I think I sorted out the JS error.

Seems like the usage of _strlen()_ and _preg_match_all()_ function is deprecated for NULL values in PHP 8.x, so I just patched adding **?? ''** in index.php:

line#160:
`
if (strlen($number ?? '')>0 || $sid) {
`
line #189
`
preg_match_all('/\[SID=(.*?)\]/i',$result[0] ?? '', $sid_data);
`
line #190
`
$sid_length = strlen($sid_data[1][0] ?? '');
`

This seems to sort out the JavaScript error and shows the calls correctly... but still not luck with the SIP ladder.

I think there are similar issues around the php code, which with new versions generate several other warning that cripple the generated page, which is the not showed correctly on the client.

The sip ladder problem seems connected to this:
![immagine](https://user-images.githubusercontent.com/25413624/224490128-a8440aca-332c-4328-84df-06fca0f2c42c.png)
working on that as well...

**Comment by niekvlessert** [comment]
Good, I was wondering about your config as well, but that it's not the issue is clear now. The install script still works pretty well. You could define the sbc_ip_used manually on the line before, but I don't think it'll get you anywhere, I think the whole SIP data object to draw is not being filled correctly. I'll have a look later.

**Comment by bbird81** [comment]
Yes, take your time obviously... I'm liking your work so I'm trying to dig it a little bit deeper.

So basically in your PHP script you try to get the ip address of the interface of the SBC in index.php in lines 303>314 and put that name into $sbc_sip_interface_ip which will be used in the drawing:

303-                                      if ($sbc_sip_interface_ip == "") {
304-                                            if (strpos($message,"Incoming SIP Message from")!=0) {
305-                                                    preg_match('/(\()(.*?)(\))/',$message, $resultt);
306-                                                    foreach ($devices_to_log_ip_interfaces[$device] as $interface_name => $ip) {
307-                                                            if ($resultt[2] == $interface_name) {
308-                                                                    $sbc_sip_interface_ip = $ip;
309:                                                                    echo "sbc_ip_used = '$ip';\n";
310-                                                            }
311-                                                    }
312-                                            }
313-                                    }
314-                            }


From my understanding you:
- filter "Incoming SIP Message from", which indicates an incoming SIP message
- try to match message number with expression '/(\()(.*?)(\))/' which seems to capture stuff between parenthesis ();
- cycle through $devices_to_log_ip_interfaces[$device], which seems to be the array written in config.php, and make a comparison between $resultt[2] (which should be the **third** occurence of something between parenthesys) and $interface_name which is the name of the interface written in config.php.

A typical message looks like this:
**[S=204063706] [SID=58ddf8:18:7579926]  (N  6805161) ---- Incoming SIP Message from 1.2.3.4:5061 to SIPInterface #4 (SIP_INTERFACE_NAME) TLS TO(#215) SocketID(597301) ---- [Time:11-03@15:22:09.000]**
So my guess is:
- $resultt[0] = N  6805161
- $resultt[1] = SIP_INT_NAME
- $resultt[2] = #215
- $resultt[3] = 597301

My wild guess was that an older format of the log put the sip interface name into the third block, but now it's in the second... This is the theory. But when I changed #307 to _if ($resultt[1] == $interface_name) {_ it still didn't work. I've tried to debug a little by adding some echo console.log here and there, what I got is that if @#303 is found true several times, but if @#304 is NEVER found true.

So I added echo "console.log('".$message."');\n"; between #303 and #304 and noticed that "Incoming SIP Message from..." are severed to something like:  **TLS TO(#215) SocketID(597301) ---- [Time:11-03@17:15:42.228]** so clearly you can never enter that condition... but I guess it also means that all parameters are read correctly as they are trimmed away.

I'm a little bit tired so I'll take a break, hope this helps you find what's going on.

PS: also got another bug cause in a custom header of PSTN provider there was some ' or " character that messed up stuff... leaving here this just to remember that headers are strings in code so it could be interpreted as a JS array/object/whatever.

**Comment by niekvlessert** [comment]
Yeah, you're correct, that's the way that part works, and yes, the syntax has changed in a few ways. I've been fixing stuff all over the place to handle that. I've got it working now, but it's not perfect; for example sometimes the syslog system of Audiocodes splits SIP messages over multiple syslog lines, my code could handle that, but now it doesn't. And it needs a long overdue clean up. ;) I will fix some more and then you can test it. Now you can also see what the message type selection is meant for.

<img width="1426" alt="image" src="https://user-images.githubusercontent.com/1013068/224550551-eda202cb-5d92-40e8-be80-ad5029804574.png">

**Comment by niekvlessert** [comment]
Ok, I pushed the modified version, let me know the results!

**Comment by bbird81** [comment]
I'll try in the next few days and let you know. Thanks!!

**Comment by bbird81** [comment]
WOW it works! I have to say I'm really impressed! 👍 

Just a couple of notes if you want to improve it a little bit more.
1. it still gave me the JS errors related to the fact that my version of PHP (8.1) has deprecation for NULL values in some functions; using the changes I proposed in my previous comment, made it work flawlessly.
2. probably more important, while the ladder is generated without problems, it doesn't render what happens with the SIP flow in the correct way... I'll explain myself with an example.

Look at this ladder (should be exactly like the logs I provided you previously):
![immagine](https://user-images.githubusercontent.com/25413624/227714511-b3b43b94-697b-405f-8895-53c5bb03f3fb.png)
Consider the INVITE going to the right towards 'x.x.x.x': it should instead go left, as it's an incoming message to an interface of AUDIOCODESLAB device.
The INVITE is structured this way:
```
SIPLadder - SIPLadder::FillVQMData - Sending CallFlow message. SesSeq=14 Direction=Tx [Time:25-03@12:10:22.495]
MEDIA - MEDIA_SYNC_REQUEST_EV: (#7)MRC -> (#7)MRC
MRC - MRC::HandleSyncRequest - SyncReason is None
MEDIA - MEDIA_SYNC_ENDED_EV: (#122)SBCController -> (#245)SBCEndPoint #012 -> (#0)SBCCall
MEDIA - MEDIA_SYNC_ENDED_EV: (#122)SBCController -> (#244)SBCEndPoint #012 -> (#1)SBCCall
Incoming - ---- Incoming SIP Message from ip.of.east.device:49831 to SIPInterface #0 (InternalVoIP) UDP TO(#0) ----
INVITE sip:1234567890@x.x.x.x:5060 SIP/2.0
Via: SIP/2.0/UDP ip.of.east.device:5060;branch=z9hG4bK5516A2655
Remote-Party-ID: "whatever" ;party=calling;screen=yes;privacy=off
From: ;tag=5D3A0308-FC8
To: "Someone" ;tag=1c1193116132
Date: Sat, 25 Mar 2023 11:10:25 GMT
Call-ID: 4430869492532023121021@x.x.x.x
Supported: 100rel,timer,resource-priority,replaces,sdp-anat
Min-SE: 1800
User-Agent: Cisco-SIPGateway/IOS-17.3.5
Allow: INVITE, OPTIONS, BYE, CANCEL, ACK, PRACK, UPDATE, REFER, SUBSCRIBE, NOTIFY, INFO, REGISTER
CSeq: 101 INVITE
Max-Forwards: 70
Timestamp: 1679742625
Contact:
Expires: 180
Allow-Events: telephone-event
Session-ID: .....
Content-Length: 0 
```
I think that the "Incoming SIP Message" should already make clear that there should be an "inbound" arrow, but just for playing it safe I've set all IPs of my SBC in the config.php as follows:
```
$devices_to_log = array (
  'AudioCodesLab' => 'mgmt.ip.address.of.device.not.equal.to.any.sip.interface',
);
$devices_to_log_ip_interfaces = array (
  'AudioCodesLab' =>
  array (
    'InternalVoIP' => 'x.x.x.x',
    'ExternalVoIP' => 'private.ip.of.west.side',
  ),
);
$none_sbc_devices = array (
  'WEST-DEVICE' => 'ip.of.west.device',
  'WBX_LG' => 'ip.of.east.device',
);
?>
```


What am I missing? Shouldn't it recognize that x.x.x.x is just another interface of AudioCodesLab device receiving an inbound message?

_TBF I have one entry for each sip interface on the device (some share ip but work on different ports), so I have multiple entries with the same ip but different names... not sure if this messes up the algorithm_

Besides this, I'd say that yours it's a wonderful piece of work! :) Many many thanks for updating it! Is there any way I can give back to you and the project? :)

**Comment by niekvlessert** [comment]
Good to see that and thanks! I'll include the NULL fixes and I think I can have a look at the weird INVITE without your data, since it's also visible in the SIP view I get from the previous tests, see a little higher. I would like to understand the stuff that is happening on the SIP & functional level though, do you maybe know? Then I don't have to recognise it from the SIP messages. ;) In my screenshot a call to 0123456789 is being setup, and the whole SIP flow is being handled with the TRYING/RINGING/PRACK/ACK etc. Then the call is online. Then after that another INVITE is happening to another number, 098765432, which is interesting, why is that?? Then that's being ACKed and then after that BYE which is normal.

**Comment by bbird81** [comment]
Yep you can use the data I gave you since it's basically the same call I'm testing with, but with redacted data.

I can explain the flow, no problem :) What you are seeing is a basic flow from PSTN to a contact center solution: west side is PSTN and east side is another session border controller which is registered to a cloud contact center.

As you said, the first flow is a basic call towards some number (in your data is 0123456789, but in the original call is the PSTN number): the call is established through the usual TRYING/RINGING/PRACK/ACK and all is good.
What you see next is a RE-INVITE which updates the so called display name (I think there's a Remote-Party-ID: header which contains the service name), so it's just labelling informations: once che call is connected, the called side wants to let you know the name of the contact center service you called, therefore sends back an INVITE without any SDP information. 

A little dissertation: this is quite common in contact centers negotiations, as you usually get "moved" from a welcome message to an IVR (or more) and then to an agent (or more): each time you're moved, you get an INVITE back which shows the calling side that you have been moved to another service, even though the media flow is never touched and remains established at the terms of the first negotiation.
This (=receive back an INVITE when the call is established) might happen also in normal calls, when you want to change something about the media; a couple of examples are a fax-over-ip-call (which is initially setup with audio only and then FoIP side adds t38 in SDP through a re-INVITE) or when the call is trasferred to another party (rtp flow must be updated with new c=..... information).

But let's get back to your flow example... so you receive this INVITE back which contains new information about the service that has been called: no SDP, so nothing changes in the RTP, just the labels. Why the different number? This is quick to explain: the called system (east) invites the calling system (west) by its **calling** number, so **098765432 should be the same number in the "from:" header of the first flow** (if not, then I've redacted something wrong, as I did a "substite all" operation :) ). This always happens when called side wants to update the call by its own initiative.
Then this is answered with a 200 OK with SDP: it could have been without SDP as well since there are no changes in the media, but that's how the west side is configured anyway (to always add SDP, typical configuration of PSTN providers), so it's just ACKed back with the same SDP used in the first flow.

So to sum it up, you are seeing a basic call from PSTN to a contact center, which has multiple services in it and when you move from one to another, it just informs of this the calling side. Then the PSTN hangs up normally.

Hope this helps understanding what's happening! If you have any other question about this flow (or SIP in general) I can try to help you, no problem ;)

**Comment by bbird81** [comment]
Just as a comparison, this what the ladder looks like in the AudioCodes tool:
![immagine](https://user-images.githubusercontent.com/25413624/227727162-11ee46e5-cc0f-48ff-a1d8-563b5680830c.png)

**Comment by niekvlessert** [comment]
Ok, fixed it, only for this case of course, but the fix should be generic, I'm interested in your findings! Make sure the IP at SIP Interface_0 in your config.php is correct. 

Btw, did you notice you can click on the arrows as well? Then it will scroll to the related SIP. :)

Maybe I should create an export method... so you can store issues. Or something else, what features are you missing?

**Comment by bbird81** [comment]
Wow that's great! I'm gonna test it in a few scenarios in my lab and let you know what I find out :)

Yep, I noticed you can click the message and it brings you to the detail of that message; I like this very much but (since you're asking for features and ideas) even more impressive would be to see the message in a "floating" square when you hover the cursor over the message, so that you don't have to leave the ladder to see a message detail... I think it improves the user experience :)

Exporting both ladder and messages in a text file would definetely be a fine addition, or even bettern an html with the ladder and all; the AudioCodes tool does this as well (but obviously only if you were connected and monitoring). Being able to archive calls and export them at will it's something definetely very very useful.

The ability to search by date/time instead of number would be nice for obvious reason. Also showing duration of call in the call list should be not too hard but nice to have (as you might spot 0 duration calls, which usually have issues).
Also being able to search for calls with "anomalies" (i.e. a call which has response code not equal to 200 or 10x in them) might be interesting and not too hard I guess. It might be a flag which can be combined with the date/time filter.

Maybe more complex to do but somehow possible and very cool, would be to show some statistics of the call (like packet loss, jitter and MOS) using CDR data. I know other tools can do that, but I'm not sure where that data is exactly taken from... I will look into this and if it's easy to find this data (like: just tie the call to CDR data using this key, the third value is packet loss, the fourth is MOS...) I will share with you details.

BTW very nice update of the README.md :D it really catches the attention now! Hope your project becomes more popular, it deserves it.
If you wish, I can spread word among the community of AudioCodes installers: you might find other people whishing to contribute... I for one I'd like to contribute in some way, let me know how you feel about it and what I could do or how could I help you.

**Comment by bbird81** [comment]
Ok, I've made a couple of tests with the updated code.

I still had to put the **?? ''** at the usual lines :)

The problem with the INVITE going east instead of west is still there; I noticed that if I change the order of the sip interfaces in config.php it works correctly:

**Rendered Correctly**
```
$devices_to_log_ip_interfaces = array (
  'AudioCodesLab' => 
  array (
    'ExternalVoIP' => 'x.x.x.x',
    'InternalVoIP' => 'y.y.y.y',
  ),
```

**Rendered Wrong**
```
$devices_to_log_ip_interfaces = array (
  'AudioCodesLab' => 
  array (
    'InternalVoIP' => 'y.y.y.y',
    'ExternalVoIP' => 'x.x.x.x',
  ),
```

It is also rendered correctly if I remove completely the reference to ExternalVoIP (I leave only InternalVoIP), which is somehow confusing.
To sum it up, it works if I put in config.php
- only InternalVoIP
- Any other interface and InternalVoIP as last one

It doesn't seem related to which interface receives the first message: I have tried to make a call in the opposite direction (Internal > External): in this scenario flow is rendered correctly regardless of which interface is declared first.

For now I have just switched the order as a workaround.

**Comment by MahoneXu** [comment]
I'm trying to use this project and I seem to be having trouble with the syslog listener part, I'll keep trying to figure it out, thank you very much for this project.

**Comment by niekvlessert** [comment]
@MahoneXu If you can't get it running, please create a new issue, with details about what your setup is, maybe if the problem has been identified I can fix the installer or at least give some pointers in there.

@bbird81 Thanks for the details, it seems my parsing logic was created a bit too hasty... I will investigate. It'll probably go wrong with multiple devices also. Did you see any other SIP scenarios going wrong? Have you been able to detect and fix some actual issue with the tool yet?

**Comment by bbird81** [comment]
Basic call flow east > west and west > east seems good (with the caveats I explained) so I'd say it's already a huge improvement; I'll check other scenarios involving multiple devices next week and let you know. Thanks for now :)
